### PR TITLE
More Hip Atomics

### PR DESCRIPTION
**Pull request by MrBurmark**
# Support more hip atomics

Refactor hip atomic implementation to support more types natively and de-duplicate implementation.

- This PR is a refactoring
- It does the following:
  - Refactors hip atomics to SFINAE based design.
  - Adds native double atomics at the request of myself

### CODE REVIEW

File: include/RAJA/policy/hip/atomic.hpp [hunk 1]
```diff
@@ -166,401 +168,300 @@ RAJA_INLINE __device__ T hip_atomic_CAS_oper(T volatile *acc, OPER &&oper)
...
+struct is_any_of<T, list<Types...>>
+  : concepts::any_of<camp::is_same<T, Types>...>
+{};
+
+template < typename T, typename TypeList >
+using enable_if_is_any_of = std::enable_if_t<is_any_of<T, TypeList>::value, T>;
+
+template < typename T, typename TypeList >
+using enable_if_is_none_of = std::enable_if_t<concepts::negate<is_any_of<T, TypeList>>::value, T>;
+
```
Comment (artv3) [inline include/RAJA/policy/hip/atomic.hpp:184]: Does any of the code above belong in camp?

Comment (trws) [reply include/RAJA/policy/hip/atomic.hpp:184]: That's probably a good helper to add with the concepts rework, I'll post up the minimal recut of that shortly and maybe we can pop it in.

**Comment by MrBurmark** [comment]
@LLNL/raja-core This is ready for review.

File: include/RAJA/policy/hip/atomic.hpp (hunk 1 above)
Comment (artv3) [reply include/RAJA/policy/hip/atomic.hpp:184]: Did this make it into camp? or are we keeping it here?

Comment (MrBurmark) [reply include/RAJA/policy/hip/atomic.hpp:184]: I didn't see it in main or the concepts refactor PR. Shall we bring this in and fix it later?

Comment (artv3) [reply include/RAJA/policy/hip/atomic.hpp:184]: deal!

**Review by artv3 — APPROVED** [review]

File: include/RAJA/config.hpp.in [hunk 2]
```diff
@@ -263,6 +263,7 @@ namespace RAJA {
 // enable device function pointers with rocm version >= 5.1
 // this used to be set to 4.3, but tests start passing with 5.1
 #define RAJA_ENABLE_HIP_INDIRECT_FUNCTION_CALL
+#define RAJA_ENABLE_HIP_DOUBLE_ATOMICADD
```
Comment (artv3) [inline include/RAJA/config.hpp.in:266]: Can we make this the default when on the supported platforms?

**Review by artv3 — CHANGES_REQUESTED** [review]

File: include/RAJA/config.hpp.in (hunk 2 above)
Comment (MrBurmark) [reply include/RAJA/config.hpp.in:266]: This is based on rocm version, not device/platform.

Comment (artv3) [reply include/RAJA/config.hpp.in:266]: Thanks I see it now, 5.1 and above right? Nice feature, I would have thought there might be a dependency on the compute arch number.

**Review by artv3 — APPROVED** [review]

**Pull request merged**